### ISSUE TITLE
Stability issues - tweak resources

### ISSUE THREAD

**Issue by dgracehmh** [zalando/zalenium#321]
**Please make sure that you provide enough information for us to help you with this issue. Thank you!**

Zalenium Image Version(s):  
3.7.1c
Docker Version:  
Docker version 17.03.1-ce, build 7392c3b/17.03.1-ce - Jenkins Slave
Docker version 17.09.0-ce, build afdb6d4 - Mac OSX
If using docker-compose, version:  
not using docker-compose
OS: 

Docker Command to start Zalenium:  
I am using ```curl -sSL https://raw.githubusercontent.com/dosel/t/i/p | bash -s start --docker-opt '--memory 512M' --chromeContainers 4 --maxDockerSeleniumContainers 5 --videoRecordingEnabled false --firefoxContainers 0 --maxTestSessions 5 --debugEnabled true```

Hi there, 
Just came across this tool yesterday and am already very impressed, it's fantastic!
I am trying to implement it but am coming across a bunch of stability issues, maybe I am missing something I can tweak or you guys can point me in the right direction.
When running locally (Mac OSx 10.11.6 - Docker version 17.09.0-ce, build afdb6d4) and also when running on a Jenkins Slave EC2 T2Medium (2 cpus, 4 Gig Mem) I am getting some stability issues. Normally Nodes seem to become unresponsive randomly - ```The following error occurred: cannot forward the request Connect to 172.17.0.6:40017 [/172.17.0.6] failed: Connection refused (Connection refused)Command duration or timeout:```
I am not running at massive scale - ```--chromeContainers 4 --maxDockerSeleniumContainers 5 --videoRecordingEnabled false --firefoxContainers 0 --maxTestSessions 5``` and running tests on 4 threads using failsafe. When I connect to the grid console I often see tests waiting in the queue for quiet a while and only 1 chrome node connected to Grid. 
On my Jenkins slave the problem seems to be worse, I am running maybe 150 tests, at the start the tests run fine but the longer the execution goes the slower/more unresponsive everything becomes, there are times Jenkins thinks the slave is offline, when i connect to Grid console it takes a long time to respond. I should have plenty of cpu's/memory to run 4 chrome nodes in parallel.
 
I am afraid I am badly managing the containers and they are killing my system (this does not seem to be the case on my MAC as Activity Monitor does not show anything alarming). I found I can ```--docker-opt '--memory 512M'```, is this for the Hub + Nodes or just Hub? Are there other settings I am missing which might affect my performance?

Sorry for the long inquiry, I really appreciate the work you guys put in to this great tool!

**Comment by diemol** [comment]
Hi @dgracehmh,

Thanks for the kind words!

Performance is something that we try to keep up with Zalenium but it is not very easy since the usage of CPU and RAM also depends on the website being tested (heavier pages make browsers use more resources and so on...).

We wrote some performance guidelines [here](https://github.com/zalando/zalenium/blob/master/docs/performance.md), which might be helpful for your case.

Now, the `--docker-opt memory 512M` flag is not really supported because it will be applied to the Zalenium container, but not to the created containers. We really have not had time to do this, so if someone is willing to give us a hand on that, we would be truly thankful. 

About your environment, in OSX it should work fairly fine, but it also depends on how much CPU and RAM you are sharing with your docker daemon. For the Jenkins machine, 2 CPUs for 4 threads should be fine, but 4GB RAM is not that much. In our environment, we tend to give 2GB per thread (we have teams running 16 threads on c3.4xlarge instances for example). But again, maybe 1GB is enough if your page is not so heavy. 

To sum it up, it mostly requires an initial tune up based on the amount of tests, threads, how heavy the page is and so on... So I would start using it sequentially, increasing to 2 threads and so on. That is what we usually need to do with every team here. 

Please, let us know how it goes after tweaking your environment, and if you want to, share your findings in our [performance link](https://github.com/zalando/zalenium/blob/master/docs/performance.md) via PR.

**Comment by dgracehmh** [comment]
Hi @diemol 

Thanks for the great feedback. I will try get some time in the future to look into passing mem config to created containers.
I will continue to play around with my setup and find optimum configuration, good to know you use 2G RAM per thread. 
I am running alot more stable now with this config ```start --chromeContainers 3 --maxDockerSeleniumContainers 3 --videoRecordingEnabled false --firefoxContainers 0``` on Jenkins (using headless chrome).
If I find anything worth sharing I will add to the performance page.
Thanks!

**Comment by diemol** [comment]
Sure, let us know how it goes, and hopefully you have some time to send us a PR!
If you need any hints on where to get started, just ping back here.

**Comment by dgracehmh** [comment]
Just some more info for the record. I've been monitoring container usage on my mac. I've assigned Docker 4cpus and 10Gig Memory, 
I rarely see a container running over 500 Mib memory..
```
CONTAINER           CPU %               MEM USAGE / LIMIT     MEM %               NET I/O             BLOCK I/O           PIDS
68e418bceed1        0.24%               367.5MiB / 9.744GiB   3.68%               1.5MB / 9.07MB      77.8kB / 336kB      101
5fe6e05481e5        0.25%               393.2MiB / 9.744GiB   3.94%               2.32MB / 20.1MB     0B / 299kB          101
cb6c6f5e2338        101.31%             110.8MiB / 9.744GiB   1.11%               6.89MB / 5.9MB      4.44MB / 246kB      26
61db756ae0e5        0.73%               466.9MiB / 9.744GiB   4.68%               60.8MB / 62.2MB     97MB / 94.2kB       87
241b6f4c5e09        15.56%              40.23MiB / 9.744GiB   0.40%               5.63kB / 2.69kB     0B / 49.2kB         10
```
